src/docker.c

```diff
@@ -0,0 +1,1754 @@
+    return 0;
+  }
+  size_t requested_bytes = size * num_members;
+  size_t actual_bytes = requested_bytes;
+  if (actual_bytes >= ctx->size) {
+    actual_bytes = ctx->size - 1;
+  }
+  memcpy(ctx->data, ptr, actual_bytes);
+  ctx->data += actual_bytes;
+  ctx->size -= actual_bytes;
+
+  // We lie about the number of bytes successfully transferred to prevent curl
+  // from returning an error to our caller. Our caller is keeping track of
+  // buffer consumption so it will independently know if the buffer filled up;
+  // the only errors it wants to hear about from curl are the more 
+  // catastrophic ones.
+  return requested_bytes;
+}
+
+/* Purpose: Make cURL call to Docker stats enpoint and retrieve response.
+ *
+ * Params: (char *) response_buffer: Pointer to memory which will hold the
+ *                                     response of the call after the function
+ *                                     returns.
+ *          (size_t) response_buffer_size: Size of the response.
+ *          (const char *) url: URL to make the cURL call to.
+ *          (const char *) socket: The socket to connect. In case of Docker
+ *                                  the socket is a UNIX socket.
+ *
+ * Returns: int: Return code. 0 for success. < 0 for failure.
+ *
+ * Notes: Using implementation with modifications from src/write_gcm.c. Removed
+ *        support for POST requests as all calls to the Docker stats API are GET
+ *        requests. Additionally this makes the call specifically to a UNIX
+ *        socket.
+ **/
+static int curl_get_json(char *response_buffer, size_t response_buffer_size,
+    const char *url, const char *socket) {
+  CURL *curl = curl_easy_init();
+  if (curl == NULL) {
+    ERROR("docker: curl_easy_init failed");
+    return DOCKER_ERROR;
+  }
+  curl_write_ctx_t write_ctx = {
+    .data = response_buffer,
+    .size = response_buffer_size
+  };
+
+  curl_easy_setopt(curl, CURLOPT_URL, url);
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, USERAGENT);
+  int status = curl_easy_setopt(curl, CURLOPT_UNIX_SOCKET_PATH, socket);
+  if (status != CURLE_OK) {
+    ERROR("docker: curl_easy_setopt() failed: %s\n",
+	  curl_easy_strerror(status));
+    curl_easy_cleanup(curl); return DOCKER_ERROR;
+  }
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &plugin_curl_write_callback);
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &write_ctx);
+  // http://stackoverflow.com/questions/9191668/error-longjmp-causes-uninitialized-stack-frame
+  curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
+  curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);	// 5 seconds.
+
+
+  int curl_result = curl_easy_perform(curl);
+  if (curl_result != CURLE_OK) {
+    ERROR("docker: curl_easy_perform() failed: %s",
+	  curl_easy_strerror(curl_result));
+    curl_easy_cleanup(curl); return DOCKER_ERROR;
+  }
+
+  long response_code;
+  curl_result = curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
+  write_ctx.data[0] = 0;
+  if (response_code >= 400) {
```

Comment:
Should we restrict this to 4XX errors? Or all all 5XX errors considered partial successes as well?